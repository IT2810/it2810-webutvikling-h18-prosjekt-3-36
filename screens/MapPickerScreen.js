import React from 'react';
import { View, StyleSheet } from 'react-native';
import { MapView, Permissions, Location } from 'expo';
import Categories from '../constants/Categories';

export default class MapPickerScreen extends React.Component {
  state = {
    // Før du har fetched første resultat fra gpsen så skal du bare være zoomer ut og se hele verdenskartet.
    mapRegion: { latitude: 0, longitude: 0, latitudeDelta: 100, longitudeDelta: 100 },
    currentLocationResult: null,
    chosenLoc: null,
  };

  componentDidMount() {
    // Fetch gps koordinater med en gang componenten har loadet.
    this._getLocationAsyncAndSetRegion();
  }

  onRegionChangeComplete = (mapRegion) => {
    this.setState({ mapRegion });
  }

  onMapViewPress = (onPressObject) => {
    const coords = onPressObject["coordinate"];
    this.setState({chosenLoc: coords});
    this.props.navigation.getParam("addLocation")(coords);
  }

  _getLocationAsyncAndSetRegion = async () => {
    // Kjøres for å zoome inn på riktig sted i verden. Første gps lokasjonen som du får.
    let { status } = await Permissions.askAsync(Permissions.LOCATION);
    if (status !== 'granted') {
      this.setState({
        currentLocationResult: 'Permission to access location was denied'
      });
    }
    // Det er en bug på android telefoner som gjør at denne ikke funker på mobiler som har "device only" på gps mode valgt i settings.
    let location = await Location.getCurrentPositionAsync({ enableHighAccuracy: false, maximumAge: 10000 });

    const lat = location["coords"]["latitude"];
    const long = location["coords"]["longitude"];
    const latDelta = 0.0922;
    const longDelta = 0.0421;
    this.setState({ mapRegion: { latitude: lat, longitude: long, latitudeDelta: latDelta, longitudeDelta: longDelta } })
  };

  render() {
    return (
      <View style={styles.container}>
        <MapView
          style={StyleSheet.absoluteFill}
          region={this.state.mapRegion}
          onRegionChangeComplete={this.onRegionChangeComplete}
          showsUserLocation={true}
          onPress={(e) => this.onMapViewPress(e.nativeEvent)}
        >
          {this.state.chosenLoc && <MapView.Marker
            coordinate={this.state.chosenLoc}
            title="Your chosen location."
            description="Click anywhere else on the map to move this marker."
          />}
        </MapView>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  }
});
