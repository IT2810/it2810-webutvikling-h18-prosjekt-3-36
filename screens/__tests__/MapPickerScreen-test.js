import React from "react";
import ShallowRenderer from 'react-test-renderer/shallow';
import MapPickerScreen from "../MapPickerScreen";
import appState from '../../store/reducers/Reducer'
import { createStore } from 'redux';

const store = createStore(appState)

it("MapPickerScreen Matches snapshot.", () => {
    const renderer = new ShallowRenderer();
    renderer.render(<MapPickerScreen title="test" onPress={() => { }} store={store} navigation={{ getParam: () => { }, setParams: () => { } }} isModalVisible={false} />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
});