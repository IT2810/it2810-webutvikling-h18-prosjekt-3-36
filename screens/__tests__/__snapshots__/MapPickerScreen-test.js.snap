// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapPickerScreen Matches snapshot. 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <MapView
    onPress={[Function]}
    onRegionChangeComplete={[Function]}
    region={
      Object {
        "latitude": 64,
        "latitudeDelta": 14,
        "longitude": 13,
        "longitudeDelta": 25,
      }
    }
    showsUserLocation={true}
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
</Component>
`;
